pilot/pkg/proxy/envoy/gateway.go

```diff
@@ -0,0 +1,122 @@
+// Copyright 2017 Istio Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package envoy
+
+import (
+	"strings"
+
+	"reflect"
+
+	meshconfig "istio.io/api/mesh/v1alpha1"
+	routing "istio.io/api/routing/v1alpha2"
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pkg/log"
+)
+
+func buildGatewayListeners(mesh *meshconfig.MeshConfig, instances []*model.ServiceInstance, config model.IstioConfigStore, ingress model.Node) Listeners {
+	gateways, err := config.List(model.Gateway.Type, model.NamespaceAll)
+	if err != nil {
+		log.Warnf("error listing gateways: %v", err)
+		return Listeners{}
+	} else if len(gateways) == 0 {
+		log.Debugf("listing gateways in namespace %q, got none", model.NamespaceAll)
+		return Listeners{}
+	}
+
+	gateway := gateways[0].Spec.(*routing.Gateway)
+	for _, spec := range gateways[1:] {
+		model.MergeGateways(gateway, spec.Spec.(*routing.Gateway))
+	}
+
+	listeners := make(Listeners, 0, len(gateway.Servers))
+	for _, server := range gateway.Servers {
+		opts := httpListenerOpt{
+			mesh:             mesh,
+			node:             ingress,
+			instances:        instances,
+			routeConfig:      nil,
+			ip:               WildcardAddress,
+			port:             int(server.Port.Number),
+			rds:              server.Port.Name,
+			useRemoteAddress: true,
+			direction:        IngressTraceOperation,
+			outboundListener: false,
+			store:            config,
+		}
+
+		switch strings.ToUpper(server.Port.Protocol) {
+		case "HTTP":
```

Comment:
HTTP or HTTP2

Reply:
Is this because we are ok with (inside of the istio mesh) making unecrypted `HTTP2` calls? I thought that all `HTTP2` calls needed to be TLS but that could have just been my understanding of how the browser would work.

Reply:
H2 is not always encrypted.. H2C also exists (http2 over plain text). Only Go enforces TLS for H2. Envoy supports plain text H2.